solr/core/src/java/org/apache/solr/core/SolrConfig.java

```diff
@@ -1162,4 +1176,19 @@ public ConfigNode get(String name) {
   public ConfigNode get(String name, Predicate<ConfigNode> test) {
     return root.get(name, test);
   }
+
+  /**
+   * Generates a String ID to represent the {@link ConfigSet}
+   *
+   * <p>Relies on the name of the ConfigSet, the name of the SolrConfig, the {@link
+   * String#hashCode()} to generate a "unique" id for the solr.xml data (including substitutions),
+   * and the version of the overlay. These 3 peices of data should combine to make a "unique"
+   * identifier for SolrConfigs, since those are ultimately all inputs to modifying the solr.xml
+   * result.
+   *
+   * @param configSetName the name of the configSet that this SolrConfig belongs to.
+   */
+  public String effectiveId(String configSetName) {
+    return configSetName + "-" + getName() + "-" + znodeVersion + "-" + rootDataHashCode;
+  }
```

Comment:
This feels off.  "Generates a String ID to represent the ConfigSet".  No... this is on SolrConfig; it should be an ID to represent the ID of this SolrConfig.  I'm skeptical configSetName even needs to be a part of this.  Sure in practice it will will be by-configSet but as a by-product of configSets having different configs.  If there is a reason ApplicationHandler needs to be by ConfigSet even if there may be identical solrconfig's (which I could imagine being so) then I think it's the caller of this method to concatenate the configSet name.  Or build the configSet's identity into an existing field to reference so that it's not passed in (I less prefer).

Reply:
Yeah the documentation was meant to say SolrConfig, not ConfigSet.


> I'm skeptical configSetName even needs to be a part of this.

I'm also skeptical, but saw no harm in adding it.

> then I think it's the caller of this method to concatenate the configSet name.

I did this originally then added the configSetName here. In general if we don't see a use case of sharing SolrConfigs between configSets, then I don't really see a need to make that a possibility. @gerlowskija can weigh in, but I don't really want to open that rats nest.

Reply:
> I'm skeptical configSetName even needs to be a part of this.

I think it should be a part of the ID.

> we don't see a use case of sharing SolrConfigs between configSets
agree

Reply:
I'm good with it being as a part of the ID so long as configSet is a field in the SolrConfig so that it's more clearly a part of its identity.  Adding as a parameter on this method is suspicious.

Reply:
> In general if we don't see a use case of sharing SolrConfigs between configSets, then I don't really see a need to make that a possibility. @gerlowskija can weigh in, but I don't really want to open that rats nest.

I didn't have a single specific reason in mind.  It just seemed prudent from an "abundance of caution" perspective.  Maybe some bug will crop up where we'd want to eliminate AH sharing, maybe a paranoid admin running a multi-tenant service would want to ensure sharing doesn't happen, etc.

That said I don't feel strongly about _where_ the configset name gets added.  I'll move it closer to the actual caching as suggested.